File: trust-tasks-ts/test/consume.test.ts
/**
 * SPEC.md §7.2 conformance tests for the TypeScript consumer pipeline.
 *
 * These deliberately mirror the test set in `trust-tasks-rs/src/consume.rs`.
 * The two reference implementations must reach the same verdict on the same
 * document; where a case exists there and not here, the languages can drift
 * apart without anything noticing.
 */

import assert from "node:assert/strict";
import { describe, it } from "node:test";

import {
  consumeInbound,
  refuse,
  rejectWithRecipient,
  respondWith,
  StaticTransport,
  UnauthenticatedTransport,
  type ConsumeOutcome,
  type ProofPolicy,
  type ProofVerifier,
  type SpecPolicy,
  type TransportContext,
  type TransportHandler,
  type TrustTaskDocument,
} from "../src/_runtime/index.js";

const ME = "did:web:maintainer.example";
const PEER = "did:web:org.example";

/** Stands in for a generated module's `SPEC`. */
const REQUIRED_SPEC: SpecPolicy = {
  typeUri: "https://trusttasks.org/spec/acl/grant/0.1",
  isBearer: false,
  isProofRequired: true,
  isRecipientRequired: true,
};

/** A spec that only RECOMMENDS a proof and does not require a recipient. */
const RELAXED_SPEC: SpecPolicy = {
  typeUri: "https://trusttasks.org/spec/acl/list/0.1",
  isBearer: false,
  isProofRequired: false,
  isRecipientRequired: false,
};

interface Payload {
  role: string;
}
interface Response {
  ok: true;
}

function doc(over: Partial<TrustTaskDocument<Payload>> = {}): TrustTaskDocument<Payload> {
  return {
    id: "req-1",
    type: REQUIRED_SPEC.typeUri,
    issuer: PEER,
    recipient: ME,
    payload: { role: "admin" },
    ...over,
  };
}

const PROOF = {
  type: "DataIntegrityProof",
  cryptosuite: "eddsa-rdfc-2022",
  verificationMethod: `${PEER}#key-1`,
  created: "2026-01-01T00:00:00Z",
  proofPurpose: "assertionMethod",
  proofValue: "z3kg",
};

const alwaysValid: ProofVerifier = { verify: () => true };
const alwaysInvalid: ProofVerifier = { verify: () => false };

const CLOCK = () => "2026-01-01T00:00:00.000Z";

function run(
  over: Partial<TrustTaskDocument<Payload>>,
  opts: {
    spec?: SpecPolicy;
    proofPolicy?: ProofPolicy;
    transport?: TransportHandler;
    handlerShouldNotRun?: boolean;
  } = {},
): Promise<ConsumeOutcome<Response>> {
  return consumeInbound<Payload, Response>({
    transport: opts.transport ?? new UnauthenticatedTransport(),
    spec: opts.spec ?? REQUIRED_SPEC,
    proofPolicy: opts.proofPolicy ?? { kind: "verify", verify: alwaysValid },
    doc: doc(over),
    myVid: ME,
    now: Date.parse("2026-01-01T00:00:00Z"),
    newErrorId: () => "err-1",
    clock: CLOCK,
    handler: (accepted) => {
      if (opts.handlerShouldNotRun) assert.fail("handler must not run");
      return respondWith<Payload, Response>(accepted, "resp-1", { ok: true }, CLOCK);
    },
  });
}

function rejectedCode(outcome: ConsumeOutcome<Response>): string {
  assert.equal(outcome.kind, "rejected", `expected rejected, got ${outcome.kind}`);
  assert.ok(outcome.kind === "rejected");
  return outcome.error.payload.code;
}

describe("§7.2 pipeline", () => {
  it("runs the handler when every check passes", async () => {
    const outcome = await run({ proof: PROOF });
    assert.equal(outcome.kind, "handled");
    assert.ok(outcome.kind === "handled");
    assert.equal(outcome.response.id, "resp-1");
    // §4.4.1 — the response carries the #response fragment...
    assert.equal(outcome.response.type, `${REQUIRED_SPEC.typeUri}#response`);
    // ...the parties swap...
    assert.equal(outcome.response.issuer, ME);
    assert.equal(outcome.response.recipient, PEER);
    // ...and §4.9 continues the thread from the request's id.
    assert.equal(outcome.response.threadId, "req-1");
  });

  it("item 4 — rejects an expired document", async () => {
    const outcome = await run({ proof: PROOF, expiresAt: "2025-01-01T00:00:00Z" }, {
      handlerShouldNotRun: true,
    });
    assert.equal(rejectedCode(outcome), "expired");
  });

  it("item 4 — treats the expiry instant itself as expired (inclusive bound)", async () => {
    const outcome = await run({ proof: PROOF, expiresAt: "2026-01-01T00:00:00Z" }, {
      handlerShouldNotRun: true,
    });
    assert.equal(rejectedCode(outcome), "expired");
  });

  it("item 5a — wrong recipient, and the error routes to the original issuer", async () => {
    const outcome = await run({ recipient: "did:web:someone-else.example", proof: PROOF }, {
      handlerShouldNotRun: true,
    });
    assert.equal(rejectedCode(outcome), "wrongRecipient");
    assert.ok(outcome.kind === "rejected");
    assert.equal(outcome.error.recipient, PEER);
  });

  it("item 5b — recipient REQUIRED but absent in-band", async () => {
    // Fires before the proof check, so it wins even though this spec also
    // requires a proof.
    const outcome = await run({ recipient: undefined }, { handlerShouldNotRun: true });
    assert.equal(rejectedCode(outcome), "malformedRequest");
  });

  it("item 7 — proof REQUIRED by the spec but absent", async () => {
    const outcome = await run({}, { handlerShouldNotRun: true });
    assert.equal(rejectedCode(outcome), "proofRequired");
  });

  it("item 7 — a RECOMMENDED spec accepts a proofless document under verify", async () => {
    // Locks in the per-spec discrimination: a regression here would make every
    // spec behave as though proof were REQUIRED.
    const outcome = await run({ proof: undefined }, { spec: RELAXED_SPEC });
    assert.equal(outcome.kind, "handled");
  });

  it("item 7 — a failing verifier maps to proofInvalid", async () => {
    const outcome = await run(
      { proof: PROOF },
      { proofPolicy: { kind: "verify", verify: alwaysInvalid }, handlerShouldNotRun: true },
    );
    assert.equal(rejectedCode(outcome), "proofInvalid");
  });

  it("item 7 — a verifier that throws is a failure, not a crash", async () => {
    const throwing: ProofVerifier = {
      verify: () => {
        throw new Error("resolver unreachable");
      },
    };
    const outcome = await run(
      { proof: PROOF },
      { proofPolicy: { kind: "verify", verify: throwing }, handlerShouldNotRun: true },
    );
    assert.equal(rejectedCode(outcome), "proofInvalid");
  });

  it("item 7 — rejectIfPresent refuses a proof-bearing document without leaking config", async () => {
    // SECURITY: a producer-supplied proof MUST NOT be silently dropped, and the
    // wire message MUST NOT describe the consumer's configuration — that would
    // let an unauthenticated probe fingerprint which endpoints lack a verifier.
    const outcome = await run(
      { proof: PROOF },
      { proofPolicy: { kind: "rejectIfPresent" }, handlerShouldNotRun: true },
    );
    assert.equal(rejectedCode(outcome), "malformedRequest");
    assert.ok(outcome.kind === "rejected");
    const msg = outcome.error.payload.message ?? "";
    assert.match(msg, /policy/);
    assert.match(msg, /§7\.2/);
    assert.doesNotMatch(msg, /verifier/);
    assert.doesNotMatch(msg, /configured/);
  });

  it("item 7 — acceptUnverified passes a proof-bearing document through", async () => {
    const outcome = await run(
      { proof: PROOF },
      { proofPolicy: { kind: "acceptUnverified" }, spec: RELAXED_SPEC },
    );
    assert.equal(outcome.kind, "handled");
  });

  it("item 8 — proof with no in-band recipient on a non-bearer spec is malformed", async () => {
    // §4.8.2 audience binding. Uses the relaxed spec so the recipient-REQUIRED
    // check of item 5b does not pre-empt the one under test.
    const outcome = await run(
      { recipient: undefined, proof: PROOF },
      {
        spec: RELAXED_SPEC,
        proofPolicy: { kind: "acceptUnverified" },
        handlerShouldNotRun: true,
      },
    );
    assert.equal(rejectedCode(outcome), "malformedRequest");
    assert.ok(outcome.kind === "rejected");
    assert.match(outcome.error.payload.message ?? "", /audience binding/);
  });

  it("item 8 — a bearer spec is exempt from audience binding", async () => {
    const bearer: SpecPolicy = { ...RELAXED_SPEC, isBearer: true };
    const outcome = await run(
      { recipient: undefined, proof: PROOF },
      { spec: bearer, proofPolicy: { kind: "acceptUnverified" } },
    );
    assert.equal(outcome.kind, "handled");
  });
});

describe("§4.8.1 party resolution", () => {
  it("fills absent in-band members from the transport", async () => {
    const transport = new StaticTransport({ issuer: PEER, recipient: ME });
    let seen: { issuer?: string; recipient?: string } | undefined;
    const outcome = await consumeInbound<Payload, Response>({
      transport,
      spec: RELAXED_SPEC,
      proofPolicy: { kind: "rejectIfPresent" },
      doc: doc({ issuer: undefined, recipient: ME }),
      myVid: ME,
      now: Date.now(),
      newErrorId: () => "err-1",
      clock: CLOCK,
      handler: (accepted, parties) => {
        seen = parties;
        return respondWith<Payload, Response>(accepted, "resp-1", { ok: true }, CLOCK);
      },
    });
    assert.equal(outcome.kind, "handled");
    assert.equal(seen?.issuer, PEER, "issuer should come from the transport");
    assert.equal(seen?.recipient, ME);
  });

  it("item 6 — in-band and transport disagreeing is an identityMismatch", async () => {
    const transport = new StaticTransport({ issuer: PEER });
    const outcome = await run(
      { issuer: "did:web:attacker.example", proof: PROOF },
      { transport, proofPolicy: { kind: "acceptUnverified" }, handlerShouldNotRun: true },
    );
    assert.equal(rejectedCode(outcome), "identityMismatch");
    assert.ok(outcome.kind === "rejected");
    // §8.1 — addressed to the transport-authenticated sender, NEVER the
    // contested in-band issuer.
    assert.equal(outcome.error.recipient, PEER);
    // §8.1/§10.4 — the message must not echo either identity.
    const msg = outcome.error.payload.message ?? "";
    assert.doesNotMatch(msg, /attacker/);
    assert.doesNotMatch(msg, /org\.example/);
  });

  it("§8.1 — identityMismatch with no transport sender is suppressed, not answered", async () => {
    // An addressed error here would be an oracle: the in-band issuer is the
    // contested identity and the transport authenticated nobody to send to.
    //
    // Reaching this needs a mismatch that survives to item 6. The document's
    // in-band recipient must equal `myVid` — otherwise item 5a rejects it as
    // `wrongRecipient` first, which routes normally and never reaches the
    // suppression path. So the disagreement is on the transport's side: it
    // reports the message as addressed elsewhere, and authenticates no sender.
    const mismatchNoSender: TransportHandler = {
      bindingUri: () => "urn:test:mismatching-no-sender",
      deriveParties: (): TransportContext => ({ recipient: "did:web:other-tenant.example" }),
    };
    const outcome = await run({}, { transport: mismatchNoSender, handlerShouldNotRun: true });
    assert.equal(outcome.kind, "suppressed");
    assert.ok(outcome.kind === "suppressed");
    assert.equal(outcome.reason.code, "identityMismatch");
  });
});

describe("handler refusals", () => {
  it("passes a handler-built error response through verbatim", async () => {
    const outcome = await consumeInbound<Payload, Response>({
      transport: new UnauthenticatedTransport(),
      spec: RELAXED_SPEC,
      proofPolicy: { kind: "rejectIfPresent" },
      doc: doc({ proof: undefined }),
      myVid: ME,
      now: Date.now(),
      newErrorId: () => "err-unused",
      clock: CLOCK,
      handler: (req) =>
        refuse(
          req,
          "err-handler",
          { code: "permissionDenied", message: "not in the maintainer vocabulary", retryable: false },
          CLOCK,
        ),
    });
    assert.equal(outcome.kind, "rejected");
    assert.ok(outcome.kind === "rejected");
    assert.equal(outcome.error.id, "err-handler");
    assert.equal(outcome.error.payload.code, "permissionDenied");
    assert.equal(outcome.error.recipient, PEER);
    assert.equal(outcome.error.issuer, ME);
  });
});

describe("§4.9.2 parentThreadId", () => {
  const PARENT = "urn:uuid:9b1d3f60-52a8-4c17-8e44-1d9c7b05f3ae";

  it("carries the parent onto the success response", () => {
    // The whole inner exchange shares one parent, so a response stays inside
    // the same enclosing exchange.
    const req = doc({ threadId: "inner-1", parentThreadId: PARENT });
    const res = respondWith<Payload, Response>(req, "resp-1", { ok: true }, CLOCK);
    assert.equal(res.parentThreadId, PARENT);
    assert.equal(res.threadId, "inner-1");
  });

  it("carries the parent onto an error response", () => {
    const req = doc({ threadId: "inner-1", parentThreadId: PARENT });
    const err = refuse(req, "err-1", { code: "taskFailed", message: "no", retryable: false }, CLOCK);
    assert.equal(err.parentThreadId, PARENT);
  });

  it("carries the parent through the full pipeline", async () => {
    const outcome = await run({ proof: PROOF, threadId: "inner-1", parentThreadId: PARENT });
    assert.ok(outcome.kind === "handled");
    assert.equal(outcome.response.parentThreadId, PARENT);
  });

  it("omits the member entirely when there is no parent", () => {
    // Not `undefined` — absent. An explicit undefined would serialise to
    // `"parentThreadId": null` under some encoders and imply a null parent.
    const res = respondWith<Payload, Response>(doc({}), "resp-1", { ok: true }, CLOCK);
    assert.ok(!("parentThreadId" in res));
    assert.equal(JSON.stringify(res).includes("parentThreadId"), false);
  });

  it("is not rejected by the pipeline on its own (§4.9.2 has no validation semantics)", async () => {
    // Even a self-referential parent — which a producer MUST NOT emit — is not
    // a consumer-side rejection: consumers MUST NOT reject on parentThreadId
    // alone.
    const outcome = await run({ proof: PROOF, threadId: "same", parentThreadId: "same" });
    assert.equal(outcome.kind, "handled");
  });
});

describe("§8.2 inResponseTo", () => {
  it("names the document the error reports on", () => {
    const err = refuse(doc({}), "err-1", { code: "proofRequired", message: "no proof", retryable: false }, CLOCK);
    assert.deepEqual(err.payload.inResponseTo, {
      typeUri: REQUIRED_SPEC.typeUri,
      id: "req-1",
    });
  });

  it("withholds the originating id under identityMismatch", () => {
    // §8.1 — the response goes to the transport-authenticated sender, not the
    // in-band issuer, and that party did not necessarily compose the document.
    // The typeUri stays: it is not identifying.
    const err = refuse(
      doc({}),
      "err-1",
      { code: "identityMismatch", message: "mismatch", retryable: false },
      CLOCK,
    );
    assert.equal(err.payload.inResponseTo?.typeUri, REQUIRED_SPEC.typeUri);
    assert.equal(err.payload.inResponseTo?.id, undefined);
  });

  it("keeps a caller-supplied value", () => {
    // The builder fills a gap; it does not overwrite a deliberate choice.
    const req = doc({});
    const err = rejectWithRecipient(
      req,
      "err-1",
      { code: "taskFailed", retryable: false, inResponseTo: { typeUri: "https://trusttasks.org/spec/acl/revoke/0.1" } },
      req.issuer,
      CLOCK,
    );
    assert.equal(err.payload.inResponseTo?.typeUri, "https://trusttasks.org/spec/acl/revoke/0.1");
    assert.equal(err.payload.inResponseTo?.id, undefined);
  });

  it("emits trust-task-error/0.4, the version whose schema has the member", () => {
    // 0.2's payload schema is additionalProperties:false, so a document
    // carrying inResponseTo would not validate as 0.2.
    const err = refuse(doc({}), "err-1", { code: "taskFailed", message: "failed", retryable: false }, CLOCK);
    assert.equal(err.type, "https://trusttasks.org/spec/trust-task-error/0.4");
  });

  it("carries through the full pipeline on a framework rejection", async () => {
    const outcome = await run({}, { handlerShouldNotRun: true }); // proofRequired
    assert.ok(outcome.kind === "rejected");
    assert.equal(outcome.error.payload.inResponseTo?.typeUri, REQUIRED_SPEC.typeUri);
    assert.equal(outcome.error.payload.inResponseTo?.id, "req-1");
  });
});

File: trust-tasks-dart/test/consume_test.dart
// SPEC.md §7.2 conformance tests for the Dart consumer pipeline.
//
// These deliberately mirror the test sets in trust-tasks-rs/src/consume.rs,
// trust-tasks-ts/test/consume.test.ts and trust-tasks-go/trusttasks. The four
// reference implementations must reach the same verdict on the same document;
// where a case exists there and not here, the languages can drift apart without
// anything noticing.

import 'package:test/test.dart';
import 'package:trust_tasks/trust_tasks.dart';

const me = 'did:web:maintainer.example';
const peer = 'did:web:org.example';

/// Stands in for a generated library's `spec`.
const requiredSpec = SpecPolicy(
  typeUri: 'https://trusttasks.org/spec/acl/grant/0.1',
  isProofRequired: true,
  isRecipientRequired: true,
);

/// A spec declaring `issuedAtRequirement: REQUIRED` (SPEC §7.3 item 17).
const freshSpec = SpecPolicy(
  typeUri: 'https://trusttasks.org/spec/acl/grant/0.1',
  isRecipientRequired: true,
  isIssuedAtRequired: true,
);

/// A spec that only RECOMMENDS a proof and does not require a recipient.
const relaxedSpec = SpecPolicy(
  typeUri: 'https://trusttasks.org/spec/acl/list/0.1',
);

class Payload {
  const Payload(this.role);
  final String role;
  Map<String, dynamic> toJson() => <String, dynamic>{'role': role};
}

class Response {
  const Response(this.ok);
  final bool ok;
  Map<String, dynamic> toJson() => <String, dynamic>{'ok': ok};
}

const testProof = Proof(
  type: 'DataIntegrityProof',
  cryptosuite: 'eddsa-rdfc-2022',
  verificationMethod: '$peer#key-1',
  created: '2026-01-01T00:00:00Z',
  proofPurpose: 'assertionMethod',
  proofValue: 'z3kg',
);

final testNow = DateTime.utc(2026, 1, 1);
String fixedClock() => '2026-01-01T00:00:00Z';
Object? payloadToJson(Payload p) => p.toJson();

TrustTaskDocument<Payload> doc({
  String? recipient = me,
  String? issuer = peer,
  String? issuedAt,
  String? expiresAt,
  String? parentThreadId,
  Proof? proof,
  Payload payload = const Payload('admin'),
}) =>
    TrustTaskDocument<Payload>(
      id: 'req-1',
      type: requiredSpec.typeUri,
      issuer: issuer,
      recipient: recipient,
      issuedAt: issuedAt,
      expiresAt: expiresAt,
      parentThreadId: parentThreadId,
      proof: proof,
      payload: payload,
    );

class _Verifier implements ProofVerifier {
  const _Verifier(this.result);
  final bool result;
  @override
  Future<bool> verify(Map<String, dynamic> doc) async => result;
}

class _ThrowingVerifier implements ProofVerifier {
  const _ThrowingVerifier();
  @override
  Future<bool> verify(Map<String, dynamic> doc) async =>
      throw StateError('resolver could not reach $peer');
}

/// Stands in for a real JSON Schema engine: refuses any payload missing a member
/// the schema's `required` names. Deliberately not a real engine — these tests
/// assert that the pipeline consults the validator and routes its verdict, not
/// that somebody else's validator is correct.
class _RequiredMembers implements PayloadValidator {
  const _RequiredMembers();
  @override
  String? validate(String schemaJson, Object? payload) {
    final required = RegExp(r'"required":\[(.*?)\]').firstMatch(schemaJson);
    if (required == null) return null;
    final names = RegExp(r'"(\w+)"')
        .allMatches(required.group(1)!)
        .map((m) => m.group(1)!)
        .toList();
    final members = (payload as Map<String, dynamic>? ?? <String, dynamic>{});
    final missing = names.where((n) => !members.containsKey(n)).toList();
    return missing.isEmpty ? null : missing.map((n) => 'missing $n').join('; ');
  }
}

class _ExplodingValidator implements PayloadValidator {
  const _ExplodingValidator();
  @override
  String? validate(String schemaJson, Object? payload) =>
      throw StateError('validator exploded');
}

Future<ConsumeOutcome<Response>> run(
  TrustTaskDocument<Payload> document, {
  SpecPolicy spec = requiredSpec,
  ProofPolicy proofPolicy = const ProofPolicy.verify(_Verifier(true)),
  PayloadPolicy payloadPolicy = const PayloadPolicy.acceptUnvalidated(),
  ConsumeChecks? checks,
  TransportHandler transport = const UnauthenticatedTransport(),
  bool handlerShouldNotRun = false,
  bool handlerReturnsNull = false,
  RejectReason? handlerRefuses,
}) =>
    consumeInbound<Payload, Response>(
      transport: transport,
      spec: spec,
      proofPolicy: proofPolicy,
      payloadPolicy: payloadPolicy,
      checks: checks ?? notConsequentialChecks(),
      doc: document,
      myVid: me,
      now: testNow,
      newErrorId: () => 'err-1',
      payloadToJson: payloadToJson,
      clock: fixedClock,
      handler: (accepted, parties) async {
        if (handlerShouldNotRun) fail('handler must not run');
        if (handlerRefuses != null) {
          throw refuse(accepted, 'err-1', handlerRefuses, clock: fixedClock);
        }
        if (handlerReturnsNull) return null;
        return respondWith<Payload, Response>(
          accepted,
          'resp-1',
          const Response(true),
          clock: fixedClock,
        );
      },
    );

String rejectedCode(ConsumeOutcome<Response> outcome) {
  expect(outcome, isA<Rejected<Response>>(), reason: 'expected a rejection');
  return (outcome as Rejected<Response>).error.payload.code;
}

void main() {
  group('§7.2 pipeline', () {
    test('runs the handler when every check passes', () async {
      final outcome = await run(doc(proof: testProof));
      expect(outcome, isA<Handled<Response>>());
      final response = (outcome as Handled<Response>).response;
      expect(response.id, 'resp-1');
      // §4.4.1 — the response carries the #response fragment...
      expect(response.type, '${requiredSpec.typeUri}#response');
      // ...the parties swap...
      expect(response.issuer, me);
      expect(response.recipient, peer);
      // ...and §4.9 continues the thread from the request's id.
      expect(response.threadId, 'req-1');
    });

    test('item 4 — rejects an expired document', () async {
      final outcome = await run(
        doc(proof: testProof, expiresAt: '2025-01-01T00:00:00Z'),
        handlerShouldNotRun: true,
      );
      expect(rejectedCode(outcome), 'expired');
    });

    test('item 4 — treats the expiry instant itself as expired', () async {
      final outcome = await run(
        doc(
          proof: testProof,
          issuedAt: '2025-12-31T23:00:00Z',
          expiresAt: '2026-01-01T00:00:00Z',
        ),
        handlerShouldNotRun: true,
      );
      expect(rejectedCode(outcome), 'expired');
    });

    test('item 4 — rejects a malformed expiresAt', () async {
      final outcome = await run(
        doc(proof: testProof, expiresAt: 'not-a-timestamp'),
        handlerShouldNotRun: true,
      );
      expect(rejectedCode(outcome), 'malformedRequest');
    });

    test('item 5a — wrong recipient routes to the original issuer', () async {
      final outcome = await run(
        doc(proof: testProof, recipient: 'did:web:someone-else.example'),
        handlerShouldNotRun: true,
      );
      expect(rejectedCode(outcome), 'wrongRecipient');
      expect((outcome as Rejected<Response>).error.recipient, peer);
    });

    test('item 5b — recipient REQUIRED but absent in-band', () async {
      final outcome = await run(
        doc(proof: testProof, recipient: null),
        handlerShouldNotRun: true,
      );
      expect(rejectedCode(outcome), 'malformedRequest');
    });

    test('item 7 — proof REQUIRED by the spec but absent', () async {
      final outcome = await run(doc(), handlerShouldNotRun: true);
      expect(rejectedCode(outcome), 'proofRequired');
    });

    test('item 7 — a RECOMMENDED spec accepts a proofless document', () async {
      final outcome = await run(doc(recipient: null), spec: relaxedSpec);
      expect(outcome, isA<Handled<Response>>());
    });

    test('§7.3 item 17 — issuedAt REQUIRED but absent', () async {
      final outcome =
          await run(doc(), spec: freshSpec, handlerShouldNotRun: true);
      expect(rejectedCode(outcome), 'malformedRequest');
      expect(
        (outcome as Rejected<Response>).error.payload.message,
        contains('§7.3 item 17'),
      );
    });

    test('§7.3 item 17 — passes once the document carries issuedAt', () async {
      final outcome =
          await run(doc(issuedAt: '2026-01-01T00:00:00Z'), spec: freshSpec);
      expect(outcome, isA<Handled<Response>>());
    });

    test('item 7 — a failing verifier maps to proofInvalid', () async {
      final outcome = await run(
        doc(proof: testProof),
        proofPolicy: const ProofPolicy.verify(_Verifier(false)),
        handlerShouldNotRun: true,
      );
      expect(rejectedCode(outcome), 'proofInvalid');
    });

    test('item 7 — a verifier that throws is a failure, not a crash', () async {
      final outcome = await run(
        doc(proof: testProof),
        proofPolicy: const ProofPolicy.verify(_ThrowingVerifier()),
        handlerShouldNotRun: true,
      );
      expect(rejectedCode(outcome), 'proofInvalid');
      // §10.4: the verifier's own text must not reach the wire.
      final message = (outcome as Rejected<Response>).error.payload.message!;
      expect(message, proofInvalidWireMessage);
      expect(message, isNot(contains('resolver')));
    });

    test('item 7 — rejectIfPresent does not leak configuration', () async {
      final outcome = await run(
        doc(proof: testProof),
        proofPolicy: const ProofPolicy.rejectIfPresent(),
        handlerShouldNotRun: true,
      );
      expect(rejectedCode(outcome), 'malformedRequest');
      expect(
        (outcome as Rejected<Response>).error.payload.message,
        proofNotAcceptedByPolicy,
      );
    });

    test('item 7 — acceptUnverified passes a proof-bearing document', () async {
      final outcome = await run(
        doc(proof: testProof),
        proofPolicy: const ProofPolicy.acceptUnverified(),
      );
      expect(outcome, isA<Handled<Response>>());
    });

    test('item 8 — proof with no recipient on a non-bearer spec', () async {
      final outcome = await run(
        doc(recipient: null, proof: testProof),
        spec: relaxedSpec,
        handlerShouldNotRun: true,
      );
      expect(rejectedCode(outcome), 'malformedRequest');
      expect(
        (outcome as Rejected<Response>).error.payload.message,
        contains('audience binding'),
      );
    });

    test('item 8 — a bearer spec is exempt from audience binding', () async {
      const bearer = SpecPolicy(
        typeUri: 'https://trusttasks.org/spec/acl/list/0.1',
        isBearer: true,
      );
      final outcome =
          await run(doc(recipient: null, proof: testProof), spec: bearer);
      expect(outcome, isA<Handled<Response>>());
    });
  });

  group('§4.8.1 party resolution', () {
    test('fills absent in-band members from the transport', () async {
      ResolvedParties? seen;
      await consumeInbound<Payload, Response>(
        transport: const StaticTransport(
          TransportContext(issuer: peer, recipient: me),
        ),
        spec: relaxedSpec,
        proofPolicy: const ProofPolicy.acceptUnverified(),
        payloadPolicy: const PayloadPolicy.acceptUnvalidated(),
        checks: notConsequentialChecks(),
        doc: doc(issuer: null, recipient: null),
        myVid: me,
        now: testNow,
        newErrorId: () => 'err-1',
        payloadToJson: payloadToJson,
        handler: (accepted, parties) async {
          seen = parties;
          return respondWith<Payload, Response>(
            accepted,
            'resp-1',
            const Response(true),
            clock: fixedClock,
          );
        },
      );
      expect(seen!.issuer, peer);
      expect(seen!.recipient, me);
    });

    test('item 6 — in-band and transport disagreeing is identityMismatch',
        () async {
      final outcome = await run(
        doc(proof: testProof),
        transport: const StaticTransport(
          TransportContext(issuer: 'did:web:impostor.example', recipient: me),
        ),
        handlerShouldNotRun: true,
      );
      expect(rejectedCode(outcome), 'identityMismatch');
      final error = (outcome as Rejected<Response>).error;
      // §8.1: addressed to the transport-authenticated sender.
      expect(error.recipient, 'did:web:impostor.example');
      // §8.1 / §10.4: the wire message names neither value.
      expect(error.payload.message, isNot(contains('impostor')));
      expect(error.payload.message, isNot(contains(peer)));
      // §8.2: the originating id is withheld under identityMismatch.
      expect(error.payload.inResponseTo!.id, isNull);
      expect(error.payload.inResponseTo!.typeUri, requiredSpec.typeUri);
    });

    test('§8.1 — identityMismatch with no transport sender is suppressed',
        () async {
      final outcome = await run(
        doc(proof: testProof),
        transport: const StaticTransport(
          TransportContext(recipient: 'did:web:other.example'),
        ),
        handlerShouldNotRun: true,
      );
      expect(outcome, isA<Suppressed<Response>>());
      expect(
        (outcome as Suppressed<Response>).reason.code,
        StandardCode.identityMismatch,
      );
    });
  });

  group('§8.2 inResponseTo and §4.9.2 parentThreadId', () {
    test('names the document the error reports on', () async {
      final outcome = await run(doc(proof: testProof, recipient: null),
          handlerShouldNotRun: true);
      final about = (outcome as Rejected<Response>).error.payload.inResponseTo!;
      expect(about.typeUri, requiredSpec.typeUri);
      expect(about.id, 'req-1');
    });

    test('error responses carry the declared error Type URI', () async {
      final outcome = await run(doc(proof: testProof, recipient: null),
          handlerShouldNotRun: true);
      final error = (outcome as Rejected<Response>).error;
      expect(error.type, trustTaskErrorTypeUri);
      expect(isErrorResponse(error), isTrue);
    });

    test('the parent thread carries onto both response kinds', () async {
      final handled =
          await run(doc(proof: testProof, parentThreadId: 'parent-1'));
      expect(
          (handled as Handled<Response>).response.parentThreadId, 'parent-1');

      final rejected = await run(
        doc(proof: testProof, recipient: null, parentThreadId: 'parent-1'),
        handlerShouldNotRun: true,
      );
      expect((rejected as Rejected<Response>).error.parentThreadId, 'parent-1');
    });

    test('omits the member entirely when there is no parent', () {
      final response = respondWith<Payload, Response>(
        doc(),
        'resp-1',
        const Response(true),
        clock: fixedClock,
      );
      expect(
        response.toJson((r) => r.toJson()).containsKey('parentThreadId'),
        isFalse,
      );
    });
  });

  group('§7.2 item 2 — payload schema validation', () {
    const withSchema = SpecPolicy(
      typeUri: 'https://trusttasks.org/spec/acl/list/0.1',
      payloadSchema: '{"type":"object","required":["role"]}',
    );
    const demandsMore = SpecPolicy(
      typeUri: 'https://trusttasks.org/spec/acl/list/0.1',
      payloadSchema: '{"type":"object","required":["role","scope"]}',
    );

    test('accepts a conforming payload', () async {
      final outcome = await run(
        doc(recipient: null),
        spec: withSchema,
        payloadPolicy: const PayloadPolicy.validate(_RequiredMembers()),
      );
      expect(outcome, isA<Handled<Response>>());
    });

    test('rejects a payload missing a REQUIRED member, before the handler',
        () async {
      final outcome = await run(
        doc(recipient: null),
        spec: demandsMore,
        payloadPolicy: const PayloadPolicy.validate(_RequiredMembers()),
        handlerShouldNotRun: true,
      );
      expect(rejectedCode(outcome), 'malformedRequest');
      expect(
        (outcome as Rejected<Response>).error.payload.message,
        contains('missing scope'),
      );
    });

    test('a validator that throws is a failure, not a crash', () async {
      final outcome = await run(
        doc(recipient: null),
        spec: withSchema,
        payloadPolicy: const PayloadPolicy.validate(_ExplodingValidator()),
        handlerShouldNotRun: true,
      );
      expect(rejectedCode(outcome), 'malformedRequest');
    });

    test('acceptUnvalidated really does skip the check', () async {
      final outcome = await run(
        doc(recipient: null),
        spec: demandsMore,
        payloadPolicy: const PayloadPolicy.acceptUnvalidated(),
      );
      expect(outcome, isA<Handled<Response>>());
    });

    test('is a no-op when the spec carries no schema', () async {
      final outcome = await run(
        doc(recipient: null),
        spec: relaxedSpec,
        payloadPolicy: const PayloadPolicy.validate(_ExplodingValidator()),
      );
      expect(outcome, isA<Handled<Response>>());
    });
  });

  group('handler outcomes', () {
    test('a handler that returns nothing yields accepted', () async {
      final outcome = await run(doc(recipient: null),
          spec: relaxedSpec, handlerReturnsNull: true);
      expect(outcome, isA<Accepted<Response>>());
    });

    test('a handler refusal passes through verbatim', () async {
      final outcome = await run(
        doc(recipient: null),
        spec: relaxedSpec,
        handlerRefuses: const RejectReason(
          code: StandardCode.permissionDenied,
          message: 'not on the approver list',
        ),
      );
      expect(rejectedCode(outcome), 'permissionDenied');
      expect((outcome as Rejected<Response>).error.recipient, peer);
    });
  });

  group('§7.1 / §7.2 unrecognized members', () {
    test('unrecognized top-level members survive a round trip', () {
      final wire = <String, dynamic>{
        'id': 'req-1',
        'type': requiredSpec.typeUri,
        'payload': <String, dynamic>{'role': 'admin'},
        'x-vendor-hint': <String, dynamic>{'keep': 'me'},
      };
      final parsed = TrustTaskDocument<Payload>.fromJson(
        wire,
        (p) => Payload((p as Map<String, dynamic>)['role'] as String),
      );
      expect(parsed.extra['x-vendor-hint'], <String, dynamic>{'keep': 'me'});
      final again = parsed.toJson(payloadToJson);
      expect(again['x-vendor-hint'], <String, dynamic>{'keep': 'me'});
    });
  });
}
